Aliases in templates

jbytheway opened the pull request:
I've been working to fix the issues around uses of typedefs / aliases in templates.  This PR replaces #707, fixes #431 and fixes #780.

However, it's not a complete solution yet.  In particular, it's still not correctly handling `libstdc++`'s `std::unordered_map`, which was the issue that originally inspired me to work on this.

So, I hope to continue to improve the solution until it does handle that case correctly.

For now, I'm posting the PR to get feedback on the general approach taken, which is as follows:

The existing code detected uses of template arguments via `VisitSubstTemplateTypeParmType` which appears in the AST whenever a template argument is used in a template instantiation.

However, when an alias of a template argument is used, no such node appears.  Instead, we get `VisitTypedefType`.

The approach of this PR is to track all aliases of template arguments in a `map`, and when `VisitTypedefType` is called on an expression using such an alias, treat it similarly to how `VisitSubstTemplateTypeParmType` used to work (the two now use a common helper function for the bulk of their functionality).

The test I've rewritten demonstrates a few cases where IWYU used to (incorrectly) suggest forward declarations were sufficient, and now treats as full-uses.

There are more cases which still fail, so (as mentioned above) I hope to understand them and improve this further.  (However, this is nevertheless an improvement on the status quo, so if you're reading this a few months after I wrote it and I haven't made further progress then it might be worth merging anyway.)

kimgr commented on iwyu.cc:
```diff
@@ -3543,6 +3572,12 @@ class InstantiatedTemplateVisitor
+  // use to the template-caller.  Sometimes those uses are hidden behind
+  // type aliases (typedefs).  This maps those aliases to the underlying
+  // template arguments.
+  map<const Type*, const Type*> template_argument_aliases_;
```
Does this need to be separate from the resugar map? I've tried a few times to get my head around the resugaring logic, without success, so I may be missing some subtlety, but they look closely related.

kimgr replied:
Hmm, yeah, that sounds weird. I would expect the resugar map to do exactly what you're attempting here + maybe some  magic for dependent types.

jbytheway replied:
I haven't tested this idea yet, but I *think* that `resugar_map` is designed to help with situations where template arguments don't match with template parameters.  For example,

```
template<typename T>
struct array_traits;

template<typename T, size_t N>
struct array_traits<T[N]> {
  static size_t size = N;
};

int foo(int) {
  char my_string[] = "...";
  return array_traits<decltype(my_string)>::size;
}
```

Here when walking the specialization of `array_traits`, there are two template parameters (`T` and `N`) corresponding to a single template argument (`decltype(my_string)`) and we need some way to match them up.  I've certainly seen some code that appears to try to deal with this sort of situation, but I've not experimented with it myself.

kimgr commented on iwyu.cc:
```diff
@@ -3118,42 +3120,68 @@ class InstantiatedTemplateVisitor
     return Base::VisitSubstTemplateTypeParmType(type);
   }
 
+  bool VisitTypedefType(clang::TypedefType* type) {
```
I have a really hard time with the template visitation. This is the `InstantiatedTemplateVisitor`, right? And I'm guessing it only traverses template instantiations. So why are all typedefs subject to resugaring in instantiated templates?

I'm thinking something like:

``` 
template<class T>
size_t GetTheSize(const T& t) {
  return sizeof(t);
}

int main() {
    float val = 0;
    return (int)GetTheSize(val);
}
```

`size_t` in the `GetTheSize` instantiation is visited here, right?

I wonder if `TemplateTypeParmTypeHelper` should be split into the part that registers the resugar entry, and the lookup? I don't know, it's all very confusing :)

jbytheway replied:
Yes, this is `InstantiatedTemplateVisitor`.

The need to track typedefs comes from examples like this:
```c++
template<typename T>
size_t GetTheSize(const T& t) {
  using myT = T;
  return sizeof(myT);
}
```
Without these changes, that template argument would be considered a forward-declare-only use.

kimgr replied:
Right, but I mean that even `size_t` is visited with this logic, even though it doesn't really need to be, right?

jbytheway replied:
I don't think `size_t` will be passed to `TemplateTypeParmTypeHelper`.  Which use of `size_t` are you thinking might cause that?

kimgr replied:
I'm really just guessing, but if `InstantiatedTemplateVisitor` traverses instantiated templates, it seems to me a use like:
```
int main() {
    float val = 0;
    return (int)GetTheSize(val);
}
```
Should visit components of `GetTheSize` and find a `typedef` in the return type. But I'll experiment a bit locally and try to bring some facts rather than theoretical worry.

I think my original trigger was that it looked like `TemplateTypeParmTypeHelper` was doing too much (the over-generic name also indicates this), and that maybe it could be split into a populate and lookup pair of methods. But let's hold off on my ideas until I've experimented a little and understand more of what's really going on :)

jbytheway replied:
Do you mean because `size_t` is itself a typedef?  It shouldn't cause a call to `TemplateTypeParmTypeHelper` because that typedef will not have been put in the `template_argument_aliases_` map.

I agree the function needs some refactoring.  I want to separate the "is is a full use" part from what is done in response to that, because currently it's a bit inconsistent.  For example, some places therein report a use at `CurrentLoc` and others at `caller_loc`.  I'm pretty sure the former is a mistake.  And only one type of uses leads to a call to `ReportExplicitInstantiations`, which also seems unlikely to be correct.

I also agree that the name isn't great.  Perhaps if I split it up better names will present themselves.

jbytheway commented:
I've pushed some further updates.  This is now feature-complete in the sense that it is sufficient to allow for use of `libstdc++`'s `std::unordered_map` to be handled correctly, which was my goal with this PR.

The major behavioural change in these new commits is that IWYU now performs more recursive template visitation.  Specifically, when an template being visited uses another template in a full-use context, passing down to it an argument that originates from the user's top-level template instantiation, it now recursively visits that other template specialization.

Currently this is unnecessarily slow, because it does so every time.  I still need to hook this into the `traversed_decls_` set so that it doesn't repeat traversals unnecessarily.

I've refactored and renamed `TemplateTypeParmTypeHelper`.  It's now two functions: `IsTemplateTypeParmUseFullUse` and `AnalyzeTemplateTypeParmUse`.

The commit history still needs cleaning up, but in terms of the overall change, this is (I think) close to what I want, so your thoughts are welcome.

kimgr commented:
Impressive! This is a lot to chew through, so it will take me a while. But just looking at the new succeeding test cases makes me really hopeful.

kimgr commented:
I can recommend https://github.com/kimgr/git-format if you want to clean up your commit series wrt clang-format.

kimgr commented on tests/cxx/typedef_in_template.cc:
```diff
@@ -7,51 +7,54 @@
-  // C++11 alias declaration, should not be an iwyu violation for T
-  using alias_type = T;
-
-  // IWYU: Pair is...*typedef_in_template-i1.h
```
I think we lost coverage for this example. The reason this was interesting was because it proves that the template is not exempt from analysis -- the instantiator of this template should not be responsible for `Pair`, it's an implementation detail of the template.

jbytheway replied:
I've restored the deleted test cases.

kimgr commented on tests/cxx/typedef_in_template.cc:
```diff
@@ -7,51 +7,54 @@
-  // The bug has instructions for how to provoke the error more obviously.
-
-  // IWYU: Class needs a declaration
-  Container<Class>::value_type vt;
```
Same here, we need to make sure that `value_type` requires `Container`, but does not force `Container` to require `Class`. I think your new testcases demonstrate this, but it's not as easily mappable back to STL-like types.

jbytheway replied:
I've restored the deleted test cases.

kimgr commented on tests/cxx/typedef_in_template.cc:
```diff
@@ -7,51 +7,54 @@
+// Verify that a full-use of an alias of a template parameter is treated as a
+// full-use of that parameter.
+template<typename T>
+struct A {
```
I think `A`, `B` and `C` could use some more evocative names to signal their role

jbytheway replied:
Renamed.

kimgr commented on tests/cxx/typedef_in_template.cc:
```diff
@@ -7,51 +7,54 @@
-#include "tests/cxx/typedef_in_template-d1.h"
-
-template<class T>
-class Container {
```
One thing I really liked with this testcase was how it showed how something STL-container-like might behave. Granted, it was too simplified to be a real example, but it told a nice story.

It might be nice to add little breadcrumb comments to describe what real-life libstdc++ or libc++ constructs your example classes map to, because that makes it easier to to see how the testcase maps to the real world, and notice when they drift apart (e.g. if we get bug reports for libstdc++ template stuff in the future, it's nice to know where to look for related test coverage)

jbytheway replied:
I've added a comment here.  There's less direct correlation than in the previous testcase, but it might help explain what's going on.

kimgr commented on run_iwyu_tests.py:
```diff
@@ -206,6 +207,7 @@ def setUp(self):
       'typedef_chain_no_follow.cc': ['.'],
       'typedef_in_template.cc': ['.'],
       'typedefs_and_resugaring.cc': ['.'],
+      'typedefs_of_template_arguments.cc': ['.'],
```
This file is not included in the patch set?

jbytheway replied:
It was added and later removed again, but I missed this line.  Good catch.

kimgr commented on iwyu_ast_util.cc:
```diff
@@ -1299,12 +1325,20 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
-    return retval;
-  const TemplateArgumentList& tpl_args
-      = tpl_decl->getTemplateInstantiationArgs();
+  const NamedDecl* decl = TypeToDeclAsWritten(tpl_spec_type);
```
I'd like to see this move down to just before its use.

jbytheway replied:
Code rearranged.

kimgr commented on iwyu_ast_util.cc:
```diff
@@ -1299,12 +1325,20 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
+  const TemplateArgument* tpl_args = tpl_spec_type->getArgs();
+  unsigned num_args = tpl_spec_type->getNumArgs();
+
+  // When this is a ClassTemplateSpecializationDecl specifically, we want to
```
And maybe move this comment up above the "default" case where we pull the args/num_args from the tpl_spec_type, to make it clearer that it all belongs together? Something like:
```
// Pull the template arguments out of the specialization type. If this is
// a ClassTemplateSpecializationDecl specifically, we want to
// get the arguments therefrom to correctly handle default arguments.
const TemplateArgument* tpl_args = tpl_spec_type->getArgs();
unsigned num_args = tpl_spec_type->getNumArgs();
const NamedDecl* decl = TypeToDeclAsWritten(tpl_spec_type);
if (const ClassTemplateSpecializationDecl* cls_tpl_decl = DynCastFrom(decl)) {
  const TemplateArgumentList& tpl_arg_list = cls_tpl_decl->getTemplateInstantiationArgs();
  tpl_args = tpl_arg_list.data();
  num_args = tpl_arg_list.size();
}
```

or something like that.

kimgr replied:
I find the loop below a little confusing when it's _not_ a `ClassTemplateSpecializationDecl` -- it seems like it's just running all types against themselves. But on the whole, I don't understand what this function does, so maybe that's something I just need to work on :)

jbytheway replied:
I don't really understand the function either, so this is something of the blind leading the blind...

kimgr commented on iwyu.cc:
```diff
@@ -3115,45 +3128,126 @@ class InstantiatedTemplateVisitor
+  // instantiated template.  Its goal is to determine whether that use should
+  // constitute a full-use by the template caller, and perform other necessary
+  // bookkeeping.
+  void AnalyzeTemplateTypeParmUse(const Type* type)
```
How about just `ReportTemplateTypeParmUse`?

Hmm, I see it's also registering `template_argument_aliases_`, I guess that's the "other necessary bookkeeping".

I wonder if you could make that more explicit by e.g. having callers pass in the `template_argument_aliases_` container? On balance, I'm not sure if it's an improvement, but worth playing around with.

jbytheway replied:
I don't think it helps.  Etiehr way it's just member functions using a member variable.  And moving it to the callers means one more reference to that variable.

kimgr commented on iwyu_ast_util.cc:
```diff
@@ -234,6 +234,14 @@ bool IsElaborationNode(const ASTNode* ast_node) {
   return elaborated_type && elaborated_type->getKeyword() != clang::ETK_None;
 }
 
+const ASTNode* MostElaboratedAncestor(const ASTNode* ast_node) {
```
No, wait. I see now that it actually _is_ the most elaborated. I think I was confused by the comment that says `Read _past_ elaborations`. I mean, it does, but ... Yeah. I think this is the right name as-is.

jbytheway commented on iwyu.cc:
```diff
@@ -2934,6 +2953,11 @@ class InstantiatedTemplateVisitor
   }
 
+  bool InvolvesKnownTemplateParam(const Type* type) const {
+    return InvolvesTypeForWhich(
```
Indeed, it seems that way.

jbytheway commented on iwyu.cc:
```diff
@@ -3075,31 +3100,19 @@ class InstantiatedTemplateVisitor
-    // do an iwyu check on this type (to see if sub-types are used).
+  // Check whether a use of a template parameter is a full use.
+  bool IsTemplateTypeParmUseFullUse(const Type* type)
+  {
```
I've used git-format to (hopefully) fix all of these and similar issues.

jbytheway commented on iwyu.cc:
```diff
@@ -3115,45 +3128,126 @@ class InstantiatedTemplateVisitor
     // we don't need to do anything: we're fine being forward-declared.
-    if (current_ast_node()->in_forward_declare_context())
-      return Base::VisitSubstTemplateTypeParmType(type);
+    if (node->in_forward_declare_context())
```
I think it shouldn't matter.  Elaboration nodes shouldn't affect forward-declaredness.

jbytheway commented on iwyu.cc:
```diff
@@ -3115,45 +3128,126 @@ class InstantiatedTemplateVisitor
+
+    // If the immediate parent node is a typedef, then register the new type as
+    // a new name for the template argument.
+    if (const TypedefNameDecl* typedef_decl
```
Fixed via git-format.

jbytheway commented on iwyu_ast_util.cc:
```diff
@@ -1171,6 +1179,24 @@ const Type* RemoveSubstTemplateTypeParm(const Type* type) {
+bool InvolvesTypeForWhich(const Type* type,
+                          std::function<bool(const Type*)> pred) {
+  type = RemoveSubstTemplateTypeParm(type);
+  if (pred(type)) return true;
```
Fixed via git-format.

jbytheway commented on iwyu_ast_util.cc:
```diff
@@ -1299,12 +1325,20 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
+
+  // When this is a ClassTemplateSpecializationDecl specifically, we want to
+  // get the arguments therefrom to correctly handle default arguments.
+  const ClassTemplateSpecializationDecl* cls_tpl_decl = DynCastFrom(decl);
```
Ah, I wondered why there were two different styles.  Thanks for clarifying.

jbytheway replied:
Code changed accordingly (one of them had to be a `dyn_cast_or_null`, and indeed they no longer fit on one line).

jbytheway commented on iwyu_ast_util.cc:
```diff
@@ -1299,12 +1325,20 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
     return retval;
 
   // Get the list of template args that apply to the decls.
-  const NamedDecl* decl = TypeToDeclForContent(tpl_spec_type);
```
Can I do that in a separate PR?

kimgr replied:
Of couse.

jbytheway commented on iwyu.cc:
```diff
@@ -1248,13 +1248,19 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   using Base::PrintableCurrentLoc;
   using Base::current_ast_node;
 
+  enum class IgnoreType {
```
Renamed as suggested.

jbytheway commented on iwyu_ast_util.cc:
```diff
@@ -1171,6 +1179,24 @@ const Type* RemoveSubstTemplateTypeParm(const Type* type) {
     return type;
 }
 
+bool InvolvesTypeForWhich(const Type* type,
```
Not sure that's better.  I feel like that's suggestive of something related to source locations.

jbytheway commented:
OK, I think I've actioned or responded to all your comments.  I've also rebased the commit history into something which is hopefully somewhat logical and comprehensible.

Reinstating the old test cases actually revealed a case which is not currently handled correctly, but it's a non-full-use being treated as a full-use, which is less bad than the opposite, so I've left it as is, with a TODO in the test case for now.

Ready for you to take another look now.

kimgr commented on tests/cxx/typedef_in_template.cc:
```diff
@@ -24,34 +25,87 @@ class Container {
+};
+
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
```
At first I found it suspicious that there was both a full- and a forward-decl-use here, but I think it kind-of makes sense. The forward-decl use is for the `using TAlias = T;`, and the full use is for the `TAlias t;`?

I guess I could argue that the forward-decl use does not belong here, because it's an internal detail of the template, but it doesn't really do any harm either, since it will be short-circuited by the full use.

kimgr replied:
After having seen a number of other template testcases during review, I think I now understand that the forward-decl is for the use of the template argument. Nothing to see here.

kimgr commented on tests/cxx/alias_template.cc:
```diff
@@ -33,6 +38,21 @@ AliasChain2<IndirectClass> aliasChain;
+// IWYU: IndirectClass is...*indirect.h
+AliasNested<IndirectClass> aliasNested;
+
+// Test the used class being nested deeper in the alias
```
I'm not sure I understand how this is different from `AliasNested` above.
The template argument is used in an evaluated context, as opposed to the unevaluated `sizeof` -- but both lead to the same IWYU result?
I wonder if the `sizeof` template should be renamed to clarify that, or something. Now it just looks like there's two identical tests after each other.

jbytheway replied:
They are roughly identical; I suspect they are redundant.  But I felt like it didn't hurt to have an extra test case.  Certainly it's important to test that both `sizeof` and regular use-as-a-variable count as full uses, and the deeper nesting was also important to test.  There was no particular reason to choose one over the other for the deeper nesting test, so I just did both.

I've renamed the templates for extra clarity and deleted a redundant comment.

jbytheway commented on iwyu_ast_util.h:
```diff
@@ -376,6 +376,10 @@ class CurrentASTNodeUpdater {
 bool IsElaborationNode(const ASTNode* ast_node);
 
+// Walk up to parents of the given node so long as each parent is an
+// elaboration node (in the sense of IsElaborationNode).
```
The case I originally actually needed to catch was the `struct X` case.  But it makes sense for the other type of elaboration also.  I've expanded the doc comment.

jbytheway commented on iwyu.cc:
```diff
@@ -3115,45 +3125,137 @@ class InstantiatedTemplateVisitor
+        VERRS(6)
+            << "Recursively traversing " << PrintableDecl(cts_decl)
+            << " which was full-used and involves a known template param\n";
+        TraverseDecl(const_cast<ClassTemplateSpecializationDecl*>(cts_decl));
```
There is precedent for doing this already.  There's a call to `TraverseDecl` in `TraverseDataAndTypeMembersOfClassHelper`.

kimgr replied:
OK, that makes my heart lighter :)

jbytheway commented on iwyu_ast_util.h:
```diff
@@ -694,6 +698,11 @@ bool IsClassType(const clang::Type* type);
 // However, vector<T> is *not* converted to vector<int>.
 const clang::Type* RemoveSubstTemplateTypeParm(const clang::Type* type);
 
+// Returns true if any type involved (recursively examinging template
```
Fixed.

jbytheway commented on iwyu.cc:
```diff
@@ -3115,45 +3125,137 @@ class InstantiatedTemplateVisitor
+
+    if (const Type* template_arg_type =
+            GetOrDefault(template_argument_aliases_, type, nullptr)) {
+      AnalyzeTemplateTypeParmUse(template_arg_type);
```
Added a comment.

jbytheway commented on tests/cxx/typedef_in_template.cc:
```diff
@@ -7,6 +7,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "direct.h"
```
Fixed.

jbytheway commented:
> > Fix another alias-in-template issue
> 
> I wish this commit message had a small example with real code. The fix is so small (good thing) that it's hard to understand where and how it has effect (bad thing) :)

You have good instincts.  This is actually the most dubious change, and isn't always correct.  If you're concerned, we can drop this commit and the remainder still works.  But I think it's a step in the right direction.  In any case, I expanded the commit message to clarify.

I think I've addressed your other comments.  Over to you again.

kimgr commented on tests/cxx/sizeof_in_template_arg.cc:
```diff
@@ -0,0 +1,41 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "direct.h"
```
`tests/cxx/direct.h`

jbytheway replied:
Fixed.

kimgr commented on iwyu.cc:
```diff
@@ -2897,10 +2903,22 @@ class InstantiatedTemplateVisitor
 
-  bool CanIgnoreType(const Type* type) const override {
-    if (!IsTypeInteresting(type) || !IsKnownTemplateParam(type))
+  bool CanIgnoreType(const Type* type, IgnoreKind ignore_type =
```
`ignore_type` -> `ignore_kind`

jbytheway replied:
Fixed.

jbytheway commented:
OK, nits fixed.

kimgr commented:
Thank you!
